### ISSUE TITLE
Regression tester regressed 833748

### ISSUE THREAD

**Issue by tamiko** [dealii/dealii#16508]
[start automated regression testsuite report]

Dear all,

this is the automated regression testsuite reporting a new regression between
- new revision: 83374885b5a04788ed5f963e33ce568bc8de5710  [regression testsuite results](https://cdash.dealii.org/index.php?project=deal.II&filtercount=1&showfilters=1&field1=revision&compare1=63&value1=833748)
- old revision: 8b581275ec3033c926999b63542f2c533cc47745  [regression testsuite results](https://cdash.dealii.org/index.php?project=deal.II&filtercount=1&showfilters=1&field1=revision&compare1=63&value1=8b5812)

Summary:
- :disappointed: new build warnings introduced for configuration [Clang-13.0.1-master-autodetection-20230116](https://cdash.dealii.org/build/842) on site tester-tos
- :disappointed: new build warnings introduced for configuration [Clang-14.0.6-master-autodetection-20230116](https://cdash.dealii.org/build/841) on site tester-tos
- :disappointed: new build warnings introduced for configuration [Clang-15.0.6-master-autodetection-20230116](https://cdash.dealii.org/build/840) on site tester-tos
- :disappointed: new build warnings introduced for configuration [Clang-16.0.6-master-autodetection-20230507](https://cdash.dealii.org/build/839) on site tester-tos
- :disappointed: new build warnings introduced for configuration [Clang-16.0.6-master-std=c++20-20230626](https://cdash.dealii.org/build/843) on site tester-tos
- :disappointed: new build warnings introduced for configuration [Clang-16.0.6-master-taskflow-20230708](https://cdash.dealii.org/build/850) on site tester-tos
- :disappointed: new build warnings introduced for configuration [Clang-16.0.6-master-unity_build-20230622](https://cdash.dealii.org/build/844) on site tester-tos

I have identified the following pull requests as possible candidates:
- #16504 (authored-by: @bangerth; merged-by: @tjhei)
- #16503 (authored-by: @bangerth; merged-by: @tjhei)
- #16502 (authored-by: @tamiko; merged-by: @tamiko)
- #16501 (authored-by: @bangerth; merged-by: @marcfehling)

Notes:
 - The regression reported above is a subset of the following full set of regressions compared to the baseline:
   - :disappointed: build warnings for configuration [Clang-13.0.1-master-autodetection-20230116](https://cdash.dealii.org/build/842) on site tester-tos
   - :disappointed: build warnings for configuration [Clang-14.0.6-master-autodetection-20230116](https://cdash.dealii.org/build/841) on site tester-tos
   - :disappointed: build warnings for configuration [Clang-15.0.6-master-autodetection-20230116](https://cdash.dealii.org/build/840) on site tester-tos
   - :disappointed: build warnings for configuration [Clang-16.0.6-master-autodetection-20230507](https://cdash.dealii.org/build/839) on site tester-tos
   - :angry: test failures for configuration [Clang-16.0.6-master-autodetection-20230507](https://cdash.dealii.org/build/839) on site tester-tos
   - :angry: test failures for configuration [Clang-16.0.6-master-sanitizer-20230622](https://cdash.dealii.org/build/849) on site tester-tos
   - :disappointed: build warnings for configuration [Clang-16.0.6-master-std=c++20-20230626](https://cdash.dealii.org/build/843) on site tester-tos
   - :disappointed: build warnings for configuration [Clang-16.0.6-master-taskflow-20230708](https://cdash.dealii.org/build/850) on site tester-tos
   - :angry: test failures for configuration [Clang-16.0.6-master-taskflow-20230708](https://cdash.dealii.org/build/850) on site tester-tos
   - :disappointed: build warnings for configuration [Clang-16.0.6-master-unity_build-20230622](https://cdash.dealii.org/build/844) on site tester-tos
   - :disappointed: build warnings for configuration [GNU-13.1.1-master-64bit_indices-20230507](https://cdash.dealii.org/build/848) on site tester-tos
   - :angry: test failures for configuration [GNU-13.1.1-master-64bit_indices-20230507](https://cdash.dealii.org/build/848) on site tester-tos
 - I will close and unpin this issue automatically once a full run is complete and compares cleanly to the baseline.
 - If closed I will reopen the issue if the testsuite run identified an additional regression compared to what I have reported so far.

[end automated regression testsuite report]

**Comment by tamiko** [comment]
@kinnewig *ping* I don't understand the logic in the following function:
```
   327     template <typename Number>                                                   
   328     Vector<Number> &                                                             
   329     Vector<Number>::operator=(const Number s)                                    
   330     {                                                                            
   331       (void)s;                                                                   
   332       Assert(s == Number(0.0),                                                   
   333              ExcMessage("Only 0 can be assigned to a vector."));                 
   334                                                                                  
   335       // As checked above, we are only allowed to use d==0.0, so pass            
   336       // a constant zero (instead of a run-time value 'd' that *happens* to      
   337       // have a zero value) to the underlying class in hopes that the compiler   
   338       // can optimize this somehow.                                              
   339       vector->putScalar(/*s=*/0.0);                                              
   340                                                                                  
   341       if (nonlocal_vector.is_null())                                             
   342         nonlocal_vector->putScalar(/*s=*/0.0);                                   
   343                                                                                  
   344       return *this;                                                              
   345     } 
```

Shouldn't it read `if(!nonlocal_vector.is_null()) ...`?

**Comment by tamiko** [comment]
The one main remaining issue are the failing tpetra tests: https://cdash.dealii.org/viewTest.php?onlyfailed&buildid=966

**Comment by kinnewig** [comment]
> ig _ping_ I don't understand the logic in the following function:
> 
> ```
>    327     template <typename Number>                                                   
>    328     Vector<Number> &                                                             
>    329     Vector<Number>::operator=(const Number s)                                    
>    330     {                                                                            
>    331       (void)s;                                                                   
>    332       Assert(s == Number(0.0),                                                   
>    333              ExcMessage("Only 0 can be assigned to a vector."));                 
>    334                                                                                  
>    335       // As checked above, we are only allowed to use d==0.0, so pass            
>    336       // a constant zero (instead of a run-time value 'd' that *happens* to      
>    337       // have a zero value) to the underlying class in hopes that the compiler   
>    338       // can optimize this somehow.                                              
>    339       vector->putScalar(/*s=*/0.0);                                              
>    340                                                                                  
>    341       if (nonlocal_vector.is_null())                                             
>    342         nonlocal_vector->putScalar(/*s=*/0.0);                                   
>    343                                                                                  
>    344       return *this;                                                              
>    345     } 
> ```
> 
> Shouldn't it read `if(!nonlocal_vector.is_null()) ...`?

Yes! This leads to a runtime error whenever this function is called and `non_local_vector.is_null == true`.
Since it tries to access `non_local_vector` even though it is not initialized.
And whenever `non_local_vector.is_null == false` the vector is not set to zero.

I addressed this in #16529.

**Comment by kinnewig** [comment]
> The one main remaining issue are the failing tpetra tests: https://cdash.dealii.org/viewTest.php?onlyfailed&buildid=966

This should be fixed by #16545.
